### PR TITLE
GUACAMOLE-1538: Refactor libguac_terminal for easier extensibility, and migrate to shared library.

### PR DESCRIPTION
**Pull request by jmuehlner**
As described in [GUACAMOLE-1538](https://issues.apache.org/jira/projects/GUACAMOLE/issues/GUACAMOLE-1538), this change modifies the `libguac_terminal` library to be a shared library, instead of being duplicated into each of the three existing terminal-based libaries (ssh, telnet, and kubernetes).

In addition, this change also refactors the way that the `guac_terminal` is created - rather than including every possible parameter into `guac_terminal_create`, there is now an `guac_terminal_options` struct, which only requires that a few required parameters be specified (client, clipboard, width, height, and dpi).

All other parameters can be set directly if needed. This should allow new parameters to be added without having to modify the call to `guac_terminal_create` for every user of this library.

### CODE REVIEW

**Review by necouchman — CHANGES_REQUESTED** [review]
One minor style nitpick, otherwise looks okay to me.

File: src/terminal/Makefile.am [hunk 1]
```diff
@@ -26,9 +26,11 @@
 AUTOMAKE_OPTIONS = foreign 
 ACLOCAL_AMFLAGS = -I m4
 
-noinst_LTLIBRARIES = libguac_terminal.la
+lib_LTLIBRARIES = libguac_terminal.la
 
-noinst_HEADERS =                 \
+libguac_terminalincdir = $(includedir)/guacamole/terminal
+
+libguac_terminalinc_HEADERS =                 \
```
[the remark this answers is not in the document]
Comment (jmuehlner) [reply src/terminal/Makefile.am]: Good catch! Updated.

**Review by necouchman — APPROVED** [review]

**Comment by mike-jumper** [comment]
I'm a bit late to this party, but:

1. We're getting some build failures when the `ldd` check runs, indicating there are some missing symbols that may cause problems for users that attempt to link against the library.
2. The `guac_common_clipboard` structure is actually an internal part of guacamole-server (the internal common library), so this library can't be used as-is. There's no way for a downstream user to create an instance of `guac_common_clipboard`. That aspect will need to either be abstracted away or additionally made public.
3. I suggest renaming to `libguac-terminal` rather than `libguac_terminal` to match convention.

File: src/terminal/terminal/terminal.h [hunk 2]
```diff
@@ -547,6 +578,88 @@ struct guac_terminal {
...
+    guac_client* client;
+
+    /**
+     * The guac_common_clipboard which will contain the current clipboard
+     * state. It is expected that this clipboard instance will be updated
+     * both internally by the terminal and externally through received
+     * clipboard instructions. This clipboard will not be automatically
+     * freed when this terminal is freed.
+     */
+    guac_common_clipboard* clipboard;
```
Comment (mike-jumper) [inline src/terminal/terminal/terminal.h:601]: My late feedback:

The actual storage and implementation of the clipboard doesn't feel like an option. `disable_copy`, etc. definitely does, but then the allocation/handling of the clipboard implementation with respect to that option should be internal to the main `guac_terminal` struct, I'd think.

**Pull request merged**